src/VisualStudio/Core/Def/Progression/GraphQueryManager.cs

```diff
@@ -21,162 +25,129 @@ namespace Microsoft.VisualStudio.LanguageServices.Implementation.Progression
+                context.OnCompleted();
             }
-
-            var solution = _workspace.CurrentSolution;
-            var tasks = liveQueries.Select(t => PopulateContextGraphAsync(solution, t.Item2, t.Item1)).ToArray();
-            var whenAllTask = Task.WhenAll(tasks);
-
-            return whenAllTask.SafeContinueWith(t => PostUpdate(solution), TaskScheduler.Default);
         }
 
-        private void PostUpdate(Solution solution)
+        private async ValueTask UpdateExistingQueriesAsync(CancellationToken cancellationToken)
         {
-            _delay = null;
+            ImmutableArray<(IGraphContext context, ImmutableArray<IGraphQuery> queries)> liveQueries;
             lock (_gate)
             {
-                // See if each context is still alive. It's possible it's already been GC'ed meaning we should stop caring about the query
-                _trackedQueries.RemoveAll(t => !IsTrackingContext(t.Item1));
-                if (_trackedQueries.IsEmpty())
+                // First, grab the set of contexts that are still live.  We'll update them below.
+                liveQueries = _trackedQueries
+                    .SelectAsArray(t => (context: t.context.GetTarget(), t.queries))
+                    .WhereAsArray(t => t.context != null)!;
+
+                // Next, clear out any context that are now no longer alive (or have been canceled).  We no longer care
+                // about these.
+                _trackedQueries = _trackedQueries.RemoveAll(t =>
                 {
-                    _workspace.WorkspaceChanged -= OnWorkspaceChanged;
-                    return;
-                }
+                    var target = t.context.GetTarget();
+                    return target is null || target.CancelToken.IsCancellationRequested;
+                });
             }
 
-            EnqueueUpdateIfSolutionIsStale(solution);
-        }
+            var solution = _workspace.CurrentSolution;
 
-        private static bool IsTrackingContext(WeakReference<IGraphContext> weakContext)
-        {
-            var context = weakContext.GetTarget();
-            return context != null && !context.CancelToken.IsCancellationRequested;
+            // Update all the live queries in parallel.
+            var tasks = liveQueries.Select(t => PopulateContextGraphAsync(solution, t.context, t.queries)).ToArray();
+            await Task.WhenAll(tasks).ConfigureAwait(false);
         }
 
         /// <summary>
         /// Populate the graph of the context with the values for the given Solution.
         /// </summary>
-        private static async Task PopulateContextGraphAsync(Solution solution, List<IGraphQuery> graphQueries, IGraphContext context)
+        private static async Task PopulateContextGraphAsync(
+            Solution solution,
+            IGraphContext context,
+            ImmutableArray<IGraphQuery> graphQueries)
         {
+            var cancellationToken = context.CancelToken;
```

Comment:
Given this method doesn't take a cancellation token but you're observing it, this probably needs an explicit catch for the OperationCancelled. otherwise it's going to escape this method and potentially end up where we don't want it to go.

Reply:
oooh.  very very good catch.  i'll doc.

Reply:
fixed.